### ISSUE TITLE
Segmentation Fault problem with Eigen::aligned_free

### ISSUE THREAD

**Issue by mamoll** [ompl/ompl#601]
**[Original report](https://bitbucket.org/ompl/ompl/issue/480) by 左方睿 (Bitbucket: [raymond zuo](https://bitbucket.org/raymond zuo), ).**

----------------------------------------

- ubuntu 16.04
- ompl 1.4
- eigen 3.3.5
- gcc 5.4

when i setup SimpleSetup.setup()

ompl::base::ProjectionMatrix::ComputeRandom call a JacobiSVD

and then have a Segmentation Fault in Eigen Memory.h line 177

```cpp
EIGEN_DEVICE_FUNC inline void aligned_free(void *ptr)
{
  #if (EIGEN_DEFAULT_ALIGN_BYTES==0) || EIGEN_MALLOC_ALREADY_ALIGNED
    std::free(ptr);    <----------here!
  #else
    handmade_aligned_free(ptr);
  #endif
}
```

it seems a Eigen aligne malloc problem

**Comment by AlejoDiaz49** [comment]
Sorry for the late reply, I thought I replied. In the end, I found that if libraries that use Eigen are not compiled using the same instructions of the CPU, let's say, it might fail. So I have

- PCL: Might have `march=native`
- OMPL: Without `march=native`
- FCL: Might have `march=native`

PCL and FCL activate this flag this way in these files:

- [PCL](https://github.com/PointCloudLibrary/pcl/blob/master/cmake/pcl_find_sse.cmake)
````ruby
if(NOT CMAKE_CROSSCOMPILING)
# Test GCC/G++ and CLANG
    if(CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUCXX OR CMAKE_COMPILER_IS_CLANG)
        include(CheckCXXCompilerFlag)
        check_cxx_compiler_flag("-march=native" HAVE_MARCH)
        if(HAVE_MARCH)
            list(APPEND SSE_FLAGS "-march=native")
        else()
            list(APPEND SSE_FLAGS "-mtune=native")
        endif()
        message(STATUS "Using CPU native flags for SSE optimization: ${SSE_FLAGS}")
    endif()
endif()
````

- [FCL](https://github.com/flexible-collision-library/fcl/blob/master/CMakeLists.txt)
````ruby
option(FCL_USE_HOST_NATIVE_ARCH "Whether FCL should use cflags from the host used to compile" OFF)
if (FCL_USE_HOST_NATIVE_ARCH)
    if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -march=native")
    else()
        message(WARNING "FCL_USE_HOST_NATIVE_ARCH is only supported in Linux. No effect.")
    endif()
endif()
````
So ass soon as I compiled OMPL with the `-march=native` flag everything seems to work just fine. The problem I had with OMPL not working with this flag was that another part of my code was using Eigen without it.

I don't know if this is an appropriate solution, so whenever I can I will try the commit you did. It seems similar to what they wrote [here](https://github.com/ethz-asl/eigen_catkin/wiki/Eigen-Memory-Issues#memory-misalignment)

**Comment by mamoll** [comment]
@AlejoDiaz49 that link at the end is very helpful. Thanks! I suspect that for Eigen::Ref<> parameters we need to explicitly pass the template parameter for memory alignment.

**Comment by jmainpri** [comment]
I just wanted to let you know that I have recompiled my code base with the system version of Eigen and now it passes my tests.

**Comment by AlejoDiaz49** [comment]
@mamoll yes, the main problems I have seen and had were solved with many of the things that appear in that link, although many of the errors are not reproducible in every PC. So basically:
 
- Always pass Eigen variables as a constant reference in the arguments of function or methods, `const Eigen&`.
- Be careful with Eigen fixed size attributes in classes, and add the`EIGEN_MAKE_ALIGNED_OPERATOR_NEW` that you already did in some places,
- The mismatch flag that appears at the end of the link, that is the problem I had with the `-march=native` flag

The rest of the things I haven't had any problems yet, although I tried to follow this list. Since OMPL is quite big there might be some places with some of these problems if they were not taken into account at first.

**Comment by AlejoDiaz49** [comment]
Hello, everyone.

It's been a while, but just to mention something that could be useful in the future.

According to the [Eigen documentation](https://eigen.tuxfamily.org/dox/group__TopicUnalignedArrayAssert.html) many of these issues are solved if you use at least Eigen 3.4 and target C++17 in the compiler.

**Comment by zkingston** [comment]
Closing this issue as a solution has been found - thank you all.
